### PR TITLE
Fix allies05a crashing

### LINKED ISSUE #16855: RA: Allies mission 5a can not proceed in the middle.
### Issue Summary

When the Spy enters the Prison to rescue Tanya, a problem arises. Please see screenshot. The log file is also attached.

#### System Information
- **Operating System:** Ubuntu 18.04
- **.NET / Mono Version:** Mono 6.0.0.313
- **OpenRA Version:** 7f4fbfc
- **Mod:** Red Alert

#### Additional Information:

- Logs

[lua.log](https://github.com/OpenRA/OpenRA/files/3438960/lua.log)

- OpenRA Replays

- Screenshots & Videos

https://www.dropbox.com/s/n8qbv8zie4qu1rw/Screenshot%20from%202019-07-28%2017-16-09.png?dl=0

https://www.dropbox.com/s/4zhwh5swurcl07i/Screenshot%20from%202019-07-28%2017-16-17.png?dl=0

https://www.dropbox.com/s/jpjy2u41006pz5p/Screenshot%20from%202019-07-28%2017-16-23.png?dl=0

### PR DESCRIPTION
**Pull request by abcdefg30**
Fixes #16855.

Testcase was infiltrating the prison without infiltrating the warfactory. This should make the script more robust against all sorts of weird things that can happen when infiltrating the prison first. Also makes warfactory infiltration a secondary objective which isn't strictly necessary but works better imho. (E.g. if you somehow managed to get the truck killed you can still win, just not by infiltrating the warfactory.)

### CODE REVIEW

File: mods/ra/maps/allies-05a/allies05a.lua [hunk 1]
```diff
@@ -202,7 +203,7 @@ FreeTanya = function()
 		Media.PlaySpeechNotification(greece, "TargetFreed")
 	end
 
-	if not SpecialCameras then
+	if not SpecialCameras and PrisonCamera then
```
Comment (pchote) [inline mods/ra/maps/allies-05a/allies05a.lua]: Isn't `if PrisonCamera then` sufficient?

File: mods/ra/maps/allies-05a/allies05a.lua [hunk 2]
```diff
@@ -319,7 +356,7 @@ InitTriggers = function()
 					greece.MarkCompletedObjective(mainObj)
 					SendReinforcements()
 
-					if SpecialCameras then
+					if SpecialCameras and PrisonCamera then
```
Comment (pchote) [inline mods/ra/maps/allies-05a/allies05a.lua]: Likewise here.

**Review by pchote — DISMISSED** [review]
:+1: to the changes here, but something is really broken for dogs to not detect spies at a range of 2 cells...

**Comment by matjaeck** [comment]
> +1 to the changes here, but something is really broken for dogs to not detect spies at a range of 2 cells...

This is because the dog needs a certain time to aim before it actually attacks. During this time, the spy can get out of range and the dog drops the target. Setting `DogJaw`'s range to 2c256 (for example) fixes this, as the dog will aim earlier and be able to attack while the target is in range.

File: mods/ra/maps/allies-05a/allies05a.lua [hunk 3]
```diff
@@ -234,16 +235,52 @@ end
...
+	end)
+
 	Trigger.OnInfiltrated(Prison, function()
+		if greece.IsObjectiveCompleted(mainObj) then
+			return
+		end
+
 		if not greece.IsObjectiveCompleted(infWarfactory) then
 			Media.DisplayMessage("Good work! But next time skip the heroics!", "Battlefield Control")
 			greece.MarkCompletedObjective(infWarfactory)
```
Comment (matjaeck) [inline mods/ra/maps/allies-05a/allies05a.lua]: Shouldn't this be `MarkFailedObjective` as you didn't and can no longer infiltrate the war factory?

**Review by ghost — DISMISSED** [review]
LGTM otherwise.

File: mods/ra/maps/allies-05a/allies05a.lua (hunk 3 above)
Comment (abcdefg30) [reply mods/ra/maps/allies-05a/allies05a.lua]: In theory yes, but in practice I disagree. The warfactory infiltration objective is only supporting, not required, so imho it makes sense to allow a flawless playthrough when you chose the harder route.

File: mods/ra/maps/allies-05a/allies05a.lua (hunk 1 above)
Comment (abcdefg30) [reply mods/ra/maps/allies-05a/allies05a.lua]: Actually not, `PrisonCamera` gets removed in two different places, depending on if `SpecialCameras` is true or not (and we don't want to attempt to remove it twice).

**Comment by abcdefg30** [comment]
Rebased and updated.

**Review by ghost — APPROVED** [review]

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8f7ce566c7f662cceb34e6959b8ff389bfbd4142

**Pull request merged**